MarkBind/markbind
# Add <page-nav-print /> to allow printing

**What is the purpose of this pull request?**




**Overview of changes:**
### Context
In #615, suggestions were made to remove the printing of pageNav, which was a valid concern as directly printing the pageNav will result in main page body be shifted slightly to the left (due to the pageNav taking up space in the right)

In https://github.com/MarkBind/markbind/issues/728#issuecomment-1206305899
I identified a use case for using page-nav as table of content
- could be important for CS2103T tP documentation migration #2072 - when printing UG/DG, page navigation will not be available and hence TOC could be quite useful

### Changes

Related to #728 but does not fully address it as this PR provides a way to print pageNav (e,g. to use as a table of content), but it is only meant to be consumed when printed. This is because at the moment, page-nav seems sufficient to replace TOC when viewing on a browser.

Mechanism:
- add event listeners to the browser such page-nav can be added/removed only for the purpose of printing
- the page-nav-print is a clone of the pageNav on the same page, less that they won't be highlighted as highlight won't work in pdf
- to determine where to insert the table of content, users can insert one or more `<page-nav-print />`

Screenshot:
> ![image](https://user-images.githubusercontent.com/41845017/212519547-c98f1a75-2477-4238-83dd-5b19a7c5ed6d.png)



**Anything you'd like to highlight/discuss:**
Some discussions done for the following points
- is supporting the `<toc />` shortcut necessary? Or is `<div class="toc" />` enough?
  - Ans: let's have a shortcut
- should the term `toc` be used? or any suggestions for a better name for the printing of page nav
  - Ans: I settled on `page-nav-print` to avoid confusion and also in line with the current page-nav related element naming
- should I mark this PR to resolve the table of content issue?
  - at the moment, it seems like there's no real use case for toc when pageNav is available (besides the print, which is addressed by this PR)
    - Ans: I think it might be good to leave the TOC issue open for further investigation

**Testing instructions:**
Manual testing:
- Follow the instructions as specified in [user guide - navigation](https://deploy-preview-2069--markbind-master.netlify.app/userguide/components/navigation#:~:text=In%20the%20page%20that%20you%20want%20to%20have%20page%20navigation%20printed%20(i.e.%20to%20serve%20as%20a%20table%20of%20content%20when%20viewed%20on%20PDFs)%2C%20use%20the%20%3Cpage%2Dnav%2Dprint%20/%3E%20component%20to%20position%20the%20pageNav%20like%20so%3A)
- can save the pdf locally and open to see that the page-nav-print works

Functional tests and unit tests added



### References
-  [Printing - Developer guides | MDN](https://developer.mozilla.org/en-US/docs/Web/Guide/Printing#using_media_queries_to_improve_layout)

**Proposed commit message: (wrap lines at 72 characters)**
Add `<page-nav-print />` to allow printing

Ability to print page-nav was removed as it disrupts the print
layout and was not useful as-is. However, printed pages do not
have access to page-nav and readers may still want an easy way
to navigate.

Let's add a `<page-nav-print />` component to allow specifying
a position to print page-nav. This component is only visible on
print.

Having the page-nav component to be printed can be useful
as it serves as table of content. It needs to be added manually as
there may not be a good default position to place it on all pages.



---

**Checklist:** :ballot_box_with_check:
